src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs

```diff
@@ -32342,9 +32284,312 @@ static void F1<T1>(T1 t1)
                 Diagnostic(ErrorCode.WRN_NullReferenceArgument, "(object)t1").WithArguments("o", "void C.F(object o)").WithLocation(8, 11));
         }
 
+        [WorkItem(27289, "https://github.com/dotnet/roslyn/issues/27289")]
+        [Fact]
+        public void NullableTInConstraint_01()
+        {
+            var source =
+@"class A { }
+class B<T> where T : T? { }
+class C<T> where T : class, T? { }
+class D<T> where T : struct, T? { }
+class E<T> where T : A, T? { }";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (4,30): error CS0701: 'T?' is not a valid constraint. A type used as a constraint must be an interface, a non-sealed class or a type parameter.
+                // class D<T> where T : struct, T? { }
+                Diagnostic(ErrorCode.ERR_BadBoundType, "T?").WithArguments("T?").WithLocation(4, 30),
+                // (2,9): error CS0454: Circular constraint dependency involving 'T' and 'T'
+                // class B<T> where T : T? { }
+                Diagnostic(ErrorCode.ERR_CircularConstraint, "T").WithArguments("T", "T").WithLocation(2, 9),
+                // (3,9): error CS0454: Circular constraint dependency involving 'T' and 'T'
+                // class C<T> where T : class, T? { }
+                Diagnostic(ErrorCode.ERR_CircularConstraint, "T").WithArguments("T", "T").WithLocation(3, 9),
+                // (5,9): error CS0454: Circular constraint dependency involving 'T' and 'T'
+                // class E<T> where T : A, T? { }
+                Diagnostic(ErrorCode.ERR_CircularConstraint, "T").WithArguments("T", "T").WithLocation(5, 9));
+        }
```

Comment:
>} [](start = 8, length = 1)

Is there any interesting scenario with nullability and possible unification on type arguments?
For instance, in `where T : I<int>, I<U>` we'll complain that `U` might unify with `int`.
But in `where T : I<int>, I<U?>` or `where T : I<int?>, I<U?>`, I think they cannot unify.

Reply:
Added tests.

---
In reply to: [197931259](https://github.com/dotnet/roslyn/pull/27525#discussion_r197931259) [](ancestors = 197931259)